From fd4c66361b98984f733ef81d51fbcd6709d1b672 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Mon, 30 Jun 2025 16:03:21 +0800
Subject: [PATCH 107/115] mtk: mac80211: support parse mtk vendor element in
 assoc rsp

Support parsing mtk vendor element in assoc rsp for station interface.
If there is mtk vendor element and its type is 0, set the vendor_mtk of
struct ieee80211_sta as true. This flag can inform the underlying driver
that the peer station is mediatek station.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 net/mac80211/ieee80211_i.h | 1 +
 net/mac80211/mlme.c        | 1 +
 net/mac80211/parse.c       | 4 ++++
 3 files changed, 6 insertions(+)

diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 56d76af2..c7135e0b 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1867,6 +1867,7 @@ struct ieee802_11_elems {
 
 	/* whether/which parse error occurred while retrieving these elements */
 	u8 parse_error;
+	bool vendor_mtk;
 };
 
 static inline struct ieee80211_local *hw_to_local(
diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 3e9fe3e2..8dcd7ba3 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -6362,6 +6362,7 @@ static bool ieee80211_assoc_success(struct ieee80211_sub_if_data *sdata,
 	sta->sta.wme = (elems->wmm_param || elems->s1g_capab) &&
 		       local->hw.queues >= IEEE80211_NUM_ACS;
 
+	sta->sta.vendor_mtk = elems->vendor_mtk;
 	err = sta_info_move_state(sta, IEEE80211_STA_ASSOC);
 	if (!err && !(ifmgd->flags & IEEE80211_STA_CONTROL_PORT))
 		err = sta_info_move_state(sta, IEEE80211_STA_AUTHORIZED);
diff --git a/net/mac80211/parse.c b/net/mac80211/parse.c
index 8847f01a..0b91b03b 100644
--- a/net/mac80211/parse.c
+++ b/net/mac80211/parse.c
@@ -422,6 +422,10 @@ _ieee802_11_parse_elems_full(struct ieee80211_elems_parse_params *params,
 					}
 				}
 			}
+			if (elen >= 4 && pos[0] == 0x00 && pos[1] == 0x0c &&
+			    pos[2] == 0xe7)
+				/* Mediatek OUI (00:0C:E7) */
+				elems->vendor_mtk = true;
 			break;
 		case WLAN_EID_RSN:
 			elems->rsn = pos;
-- 
2.45.2

